multi: Add decred.

JoeGruffins opened the pull request:
This is not complete but wanted to get feedback. Trying to add decred here with the intent of adding it to BTCpay Server. I appreciate any feedback and will do whatever it takes to get decred added. Thanks for your time.

JoeGruffins commented:
Rebasing and squashing and adding more. Please have a brief look though...

pubpete commented:
Any update on this? Would be a great addition for Decred as Cake Wallet is the go to mobile wallet for the community.

3pillars commented:
Let's getting done guys.

JoeGruffins commented:
We are working on it. We were told somewhere that tests should not be changed. The pr currently does that. Removing those and some more changes still incoming.

JoeGruffins committed:
tests: Add Decred unit tests.

Add 20 unit tests covering:
- Network loading and lookup for mainnet/testnet/regtest
- P2PKH and P2SH address parsing (mainnet and testnet)
- Address generation and roundtrip on all networks
- WIF private key parsing and roundtrip (mainnet and testnet)
- Blake256 hash verification (Decred vs Bitcoin address differ)
- Genesis block parsing and roundtrip
- Block header serialization (180-byte format)
- Transaction serialization/deserialization roundtrip
- Prefix-only tx hashing verification
- ConsensusFactory type checks and RPC properties
- Network property validation (ports, factory type)

JoeGruffins committed:
multi: Add decred.

Add Decred (DCR) as an altcoin to NBitcoin with support for:

- Decred transaction serialization (prefix/witness split, expiry, tree)
- Block header with stake fields (StakeRoot, VoteBits, Voters, etc.)
- Blake256/Blake3 hashing for PoW, checksums, and merkle trees
- DCP0005 block header commitments (combined regular+stake merkle root)
- DCP0011 ASERT difficulty algorithm with Blake3 PoW
- DecredGetInitStateBehavior for P2P handshake
- Base58 with Blake256 checksums and WIF private key sig scheme byte
- RIPEMD160(Blake256) for Hash160 (address derivation)
- Custom RPC handling via ConsensusFactory virtual methods
- Proper GetSignatureHash with SigHashNone, SigHashSingle,
  SigHashAnyOneCanPay support per dcrd spec
- Decred-specific RPC response parsing (getbalance, gettxoutsetinfo,
  getpeerinfo, getrawtransaction, getblock)
- Network definitions for mainnet, testnet, and simnet/regtest

All Decred-specific code is fully isolated in NBitcoin.Altcoins/Decred.cs.
Core NBitcoin changes are limited to altcoin-agnostic extension points:
- NetworkStringParser.CreateP2PKH(PubKey, Network) for custom Hash160
- ConsensusFactory RPC virtual methods for altcoin RPC customization
- Block.GetMerkleRoot() made virtual
- Consensus.GetWorkRequired() hook for custom difficulty algorithms

NicolasDorier commented:
mmh, this is a shame. Yes, the checksum isn't changeable, and changing it would require such a big refactoring in NBitcoin and NBitcoin's user code that we can't really do this.

Implementing Decred as a plugin is probably a good idea, the downside is that you won't depend on NBXplorer so features of BTCPay Server that we maintain out of the box such as integrated wallet, won't be available. (You can develop your own stuff though)
